dll/win32/shlwapi/path.c

```diff
@@ -1353,7 +1353,7 @@ BOOL WINAPI PathFindOnPathExW(LPWSTR lpszFile,LPCWSTR *lppszOtherDirs,DWORD dwWh
 
   TRACE("(%s,%p,%08x)\n", debugstr_w(lpszFile), lppszOtherDirs, dwWhich);
 
-  if (!lpszFile || !PathIsFileSpecW(lpszFile))
+  if (!PathIsFileSpecW(lpszFile) || !lpszFile)
```

Comment:
Why did you reorder this?

Reply:
@learn-more  We take advantage of the fact that the statements inside the if are executed from right to left.
First it checks if the pointer is valid (not NULL) if TRUE executes the left instruction.
This prevents the pointer from being dereferenced if it is null.

Reply:
elegant isn't it?

Reply:
wouldn't it be much better than to turn that from an OR condition into an AND condition where we do have the guarantee that stuff is evaluated from left to right? that would be much more readable.
You remember that each condition of the kind
(!A) OR (!B)
can be written logically equivalent also as
!(A AND B).

Reply:
> @learn-more We take advantage of the fact that the statements inside the if are executed from right to left. First it checks if the pointer is valid (not NULL) if TRUE executes the left instruction. This prevents the pointer from being dereferenced inside PathIsFileSpecW if it is null.

Except the statements are executed left to right.

Reply:
and `PathIsFileSpecW` returns `FALSE` for a nullptr, so it wouldn't make any difference anyway.

Reply:
learn-more: "Except the statements are executed left to right."
I always thought that there is no guarantee at all about the order of evaluations for OR conditions.
Only AND conditions are guaranteed to be evaluated from left to right iirc. Therefore my proposal for the change.

learn-more: "and PathIsFileSpecW returns FALSE for a nullptr, so it wouldn't make any difference anyway."
I have not double-checked that, and therefore can neither confirm nor want to put that into question.

Reply:
Given the following answer on stackoverflow https://stackoverflow.com/a/2456415 it seems like Mark Jansen is right and both OR and AND should guarantee evaluation from left to right.
But that doesn't reflect what zecarlos claimed (that exchanging the conditions in the OR would make a difference), and it also contradicts what I do remember personally for some reason.

Reply:
I don't understand. I made this change because I had a failed test due to an exception (dereferencing a NULL pointer).
But now I see that this exception does not occur???